src/game/AI/ScriptDevAI/scripts/outland/hellfire_citadel/shattered_halls/shattered_hallsScripts.cpp

```diff
@@ -270,6 +269,251 @@ struct npc_shattered_hand_centurion : public CombatAI
+            // When Legionnaire 001 is dead change worldstate to false, heathen/savage group around him cant respawn anymore
+            m_creature->GetMap()->GetVariableManager().SetVariable(WORLD_STATE_LEGIONNAIRE_001, 0);
+        }
+    }
+
+    void ReceiveAIEvent(AIEventType eventType, Unit* /*sender*/, Unit* /*invoker*/, uint32 /*miscValue*/) override
+    {
+        if (eventType == AI_EVENT_JUST_DIED)
+            ResetTimer(LEGIONNAIRE_CALL_FOR_REINFORCEMENTS, 0u);
+    }
+
+    void SummonedCreatureJustDied(Creature* summoned) override
+    {
+        // There can always be just one reinforcement up, when summoned dies legionnaire can spawn a new one after ~5 seconds cooldown
+        if (m_reinfCD)
+            ResetTimer(LEGIONNAIRE_REINF_CD, urand(10000, 15000));
+    }
+
+    // also reset timer when summoned despawns 
+    void SummonedCreatureDespawn(Creature* /*summoned*/) override
+    {
+        if (m_reinfCD)
+            ResetTimer(LEGIONNAIRE_REINF_CD, urand(10000, 15000));
+    }
+
+    void SummonedMovementInform(Creature* summoned, uint32 /*motionType*/, uint32 pointId) override
+    {
+        // When last waypoint reached, search for players.
+        if (pointId == 100)
+        {
+            summoned->GetMotionMaster()->MoveIdle();
+            summoned->SetInCombatWithZone();            
+        }
+    }
+
+    void CallForReinforcements()
+    {
+        // reinforcement can get spawned even if legionnaire is outfight and has a cooldown between 10 and 15 seconds, but only one can be up
+        if (!m_reinfCD)
+        {
+            if (m_creature->HasStringId(SECOND_LEGIONNAIRE_STRING) || m_creature->HasStringId(THIRD_LEGIONNAIRE_STRING) || m_creature->HasStringId(FOURTH_LEGIONNAIRE_STRING) || m_creature->HasStringId(FIFTH_LEGIONNAIRE_STRING))
+            {
+                if (Creature* felorc = m_creature->SummonCreature(MOB_FEL_ORC, FelOrcSpawnCoords[legionnaireGuid][0], FelOrcSpawnCoords[legionnaireGuid][1], FelOrcSpawnCoords[legionnaireGuid][2], FelOrcSpawnCoords[legionnaireGuid][3], TEMPSPAWN_TIMED_OOC_OR_DEAD_DESPAWN, urand(20000, 25000), true, true))
+                {
+                    felorc->GetMotionMaster()->MoveWaypoint(legionnaireGuid, 0, 0, 0, FORCED_MOVEMENT_RUN);
+                    felorc->SetCanCallForAssistance(false);
+                    felorc->SetCanCheckForHelp(false);
+                    DoBroadcastText(aRandomReinf[urand(0, 6)], m_creature);
+                }
+            }
+            m_reinfCD = true;
+        }
+
+        // Legionnaire 006
+        if (m_creature->HasStringId(SIX_LEGIONNAIRE_STRING))
+        {
+            // there are 4 sleeping npcs around him, if one of his group members dies he will call for one of the sleeping creatures to get up and join the fight
+            // this doesnt have a cd, if all 4 npcs with sleeping aura are up, nothing more happens
+            auto m_sleepingReinf = m_creature->GetMap()->GetCreatures(SLEEPING_REINF_STRING);
```

Comment:
I need to figure out an error system for this. Likely in the method itself. What can happen is that the string id isnt yet in db and this will crash.